### PR TITLE
AWS Connection Implementation

### PR DESCRIPTION
**Pull request by moulimukherjee**
### Motivation
Implementation for https://github.com/MaterializeInc/materialize/issues/23055
Depends upon https://github.com/MaterializeInc/cloud/pull/8224

### Tips for reviewer
Comments inline. 

Split out a PR to just accept the new environmentd cli arg https://github.com/MaterializeInc/materialize/pull/23626, in case the current PR is delayed. Will rebase once that's merged.

This PR is making backwards incompatible changes to the `CREATE CONNECTION` SQL for AWS connection. This has been put behind a feature flag and nobody's using it yet, so it should be fine.

Currently changes are still required on the cloud's side to pass a new AWS role arn to environmentd https://github.com/MaterializeInc/cloud/pull/8224. I am trying to get the materialize changes in before I go on a vacation next week.

I will create a separate docs PR when this goes to private preview.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  - [Rendered design doc](https://github.com/MaterializeInc/materialize/blob/2141c409057c1b2047fda423e3b6279e66510096/doc/developer/design/20231110_aws_connections.md)
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - No user facing changes yet. This feature has been put behind a feature flag, and nobody's using it.

cc @benesch @jubrad @alex-hunt-materialize @hlburak

### CODE REVIEW

File: src/environmentd/src/bin/environmentd/main.rs
Comment (moulimukherjee) [inline src/environmentd/src/bin/environmentd/main.rs:563]: @jubrad Using `AWS_EXTERNAL_CONNECTION_ROLE` to pass the new role to assume. Lmk if you would prefer something else.

File: src/adapter/src/catalog/config.rs
Comment (jkosh44) [inline src/adapter/src/catalog/config.rs:237]: Is there a GH issue to link this to?

File: src/sql/src/plan/statement/ddl/connection.rs
Comment (jkosh44) [inline src/sql/src/plan/statement/ddl/connection.rs:225]: We tend to prefer structured errors to `sql_bail!` and `sql_err!`.

**Review by jkosh44 — APPROVED** [review]
Adapter/SQL changes LGTM.

**Comment by benesch** [comment]
Gonna take a look at this by EOD!

File: src/sql/src/plan/statement/ddl/connection.rs
Comment (moulimukherjee) [reply src/sql/src/plan/statement/ddl/connection.rs]: Done!

Comment (moulimukherjee) [reply src/sql/src/plan/statement/ddl/connection.rs]: I will do a follow up PR to make use of the new errors at other places if possible.

File: src/storage-types/src/connections/aws.rs
Comment (benesch) [inline src/storage-types/src/connections/aws.rs:215]: As mentioned above, this should change to take a `&ConnectionContext` as input. You can pull the secrets reader out of that, and also the ARN of the intermediary role.

File: src/storage-types/src/connections/aws.rs
Comment (benesch) [inline src/storage-types/src/connections/aws.rs:35]: As mentioned below, I think this type should disappear, with fields moving to the `ConnectionContext`...

File: src/storage-types/src/connections/aws.rs
Comment (benesch) [inline src/storage-types/src/connections/aws.rs:65]: ...and these can become methods on the `AwsConfig`. I'd drop the `aws` from `get_aws_example_trust_policy` since it'll be implied by `AwsConfig`.

File: src/storage-types/src/connections/aws.rs
Comment (benesch) [inline src/storage-types/src/connections/aws.rs:38]: I'd be explicit and swap `aws_external_id_suffix` for `connection_id: GlobalId`.

File: src/storage-types/src/connections/aws.rs
Comment (benesch) [inline src/storage-types/src/connections/aws.rs:211]:
```suggestion
    pub async fn load_sdk_config(
```

Recommend sticking with "load" in the name to indicate this function does work (vs being a simple accessor).

File: src/storage-types/src/connections/aws.rs
Comment (benesch) [inline src/storage-types/src/connections/aws.rs:214]: As above, recommend being explicit here that this is `connection_id: GlobalId`.

Also, the fact that this parameter is optional is a recipe for misuse. Recommend the following:

  * Split the implementation of this method to new methods `AwsCredentials::load_sdk_config` and `AwsAssumeRole::load_sdk_config`.
  * Teach this method to delegate to the new methods.
  * Replace `external_id_suffix: Option<String>` with `connection_id: GlobalId` to force all callers to do the good thing.
  * Add a `AwsAssumeRole::dangerously_load_sdk_config_without_external_id` function that shares code with `AwsAssumeRole` via a helper method or somesuch. This method can be used by `validate_connection` to test the case where the external ID is not set, but its unwieldy and scary name should prevent it from being inappropriately used in other contexts.

File: src/storage-types/src/connections/aws.rs
Comment (benesch) [inline src/storage-types/src/connections/aws.rs:317]: These errors will be user facing. Are they descriptive enough to stand on their own, or do we need to massage them?

File: src/storage-types/src/connections/aws.rs
Comment (benesch) [inline src/storage-types/src/connections/aws.rs:337]: This will be a user facing error and so should follow the error message style guide: https://github.com/MaterializeInc/materialize/blob/main/doc/developer/style.md#error-message-style. In particular a hint here would be really useful, since I imagine this will be quite a confusing error message for most users.

File: src/storage-types/src/connections/aws.rs
Comment (benesch) [inline src/storage-types/src/connections/aws.rs:279]: I'm not sure this is the right way to do role chaining. Unfortunately I can't find an example online, but I think you want to stack two `AssumeRoleProvider`s together somehow. Otherwise, when the credentials that got "hardcoded" into `credentials` expire, the SDK credentials will be permanently expired—instead of the `AssumeRoleProvider` just fetching new credentials. I think!

File: src/storage-types/src/connections/aws.rs
Comment (benesch) [inline src/storage-types/src/connections/aws.rs:239]: I know these `unwrap()`s were already here, but now that this is becoming production code they should turn into errors. They can say "internal error" in the message, because they _should_ never fail, but they can in rare cases fail (they involve a network call) and so we should be careful not to panic `envd` if they fail. A few more such instances below.

File: test/testdrive/aws-connection.td
Comment (benesch) [inline test/testdrive/aws-connection.td:24]: This would be a good moment to use the syntax that lets you include/assert column headers in the result: https://github.com/MaterializeInc/materialize/blob/7bdcf673c18bfa14ea14120dd76a01e712c74691/test/testdrive/array.td#L14-L15

That should make it easier to follow which columns are which! Also you don't need to wrap each column in quotes. Just the first one (because it starts with $).

File: test/testdrive/aws-connection.td
Comment (benesch) [inline test/testdrive/aws-connection.td:36]: Ditto.

File: src/sql/src/catalog.rs
Comment (benesch) [inline src/sql/src/catalog.rs:1138]: I think you should be able to avoid this by not storing the AWS info in the connection config; see future comment.

File: src/sql/src/catalog.rs
Comment (benesch) [inline src/sql/src/catalog.rs:284]: I think you should be able to avoid this by not storing the AWS info in the connection config; see future comment.

File: src/sql/src/plan/statement/ddl/connection.rs
Comment (benesch) [inline src/sql/src/plan/statement/ddl/connection.rs:228]: See future comment about not storing `mz_principal` in the connection config; if you do so you shouldn't need to pull anything out of the catalog or add the new `MissingState` variant.

File: src/adapter/src/catalog/config.rs
Comment (benesch) [inline src/adapter/src/catalog/config.rs:227]: See later comment about moving the `mz_principal` to `ConnectionContext`, but I don't think the `aws_external_connection_role` wants/needs to be here!

File: test/testdrive/mzcompose.py
Comment (benesch) [inline test/testdrive/mzcompose.py:98]: We try to keep testdrive runnable without mzcompose—i.e., via `cargo run --bin testdrive`. This will break that ability, as aws-accounts.td will depend on these variables being set. 

It would probably be best to move the AWS connection tests to a new mzcompose composition that starts localstack. Look at test/secrets-aws-secrets-manager for inspiration. @MaterializeInc/testing can help. Possibly the right thing to do is to reuse the existing composition in test/secrets-aws-secrets-manager and simply add a new workflow to the file.

File: src/adapter/src/catalog/builtin_table_updates.rs
Comment (benesch) [reply src/adapter/src/catalog/builtin_table_updates.rs]: Fixed in https://github.com/MaterializeInc/materialize/pull/23642.

File: src/storage-types/src/connections/aws.rs
Comment (moulimukherjee) [reply src/storage-types/src/connections/aws.rs:279]: So, I saw the following in the docs https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_terms-and-concepts.html
> Role chaining is when you use a role to assume a second role through the AWS CLI or API. For example, RoleA has permission to assume RoleB. You can enable User1 to assume RoleA by using their long-term user credentials in the AssumeRole API operation. This returns RoleA **short-term credentials**. With role chaining, you can use RoleA's short-term credentials to enable User1 to assume RoleB.

Let me double check the API. I found a java example https://exceptionly.com/2022/03/09/aws-iam-role-chaining-in-java-dynamodb-s3/ which does something similar to what you mentioned. I think rust should be similar.

File: src/adapter/src/catalog/builtin_table_updates.rs
Comment (moulimukherjee) [reply src/adapter/src/catalog/builtin_table_updates.rs:792]: I had tried this and other variations, but I was getting borrow checker errors and the policy field was the only reason that I couldn't use `pack_slice`. Let me give it a try again, will paste the error otherwise.

Comment (moulimukherjee) [reply src/adapter/src/catalog/builtin_table_updates.rs]: This is the error
```
 --> src/adapter/src/catalog/builtin_table_updates.rs:785:42
    |
785 |             example_trust_policy.map(|p| p.into_row().into_element()),
    |                                          ------------^^^^^^^^^^^^^^^
    |                                          |
    |                                          returns a value referencing data owned by the current function
    |                                          temporary value created here

```

File: test/testdrive/mzcompose.py
Comment (moulimukherjee) [reply test/testdrive/mzcompose.py:98]: I will move it to test/secrets-aws-secrets-manager. I was not sure whether to create a completely new mzcompose or use something existing.

File: src/storage-types/src/connections/aws.rs
Comment (moulimukherjee) [reply src/storage-types/src/connections/aws.rs:279]: (I was not able to test out the assume role flow since the new role changes were not in place)

File: src/sql/src/catalog.rs
Comment (moulimukherjee) [reply src/sql/src/catalog.rs]: done

File: src/sql/src/plan/statement/ddl/connection.rs
Comment (moulimukherjee) [reply src/sql/src/plan/statement/ddl/connection.rs]: updated

File: src/adapter/src/catalog/config.rs
Comment (moulimukherjee) [reply src/adapter/src/catalog/config.rs]: done

File: src/adapter/src/catalog.rs
Comment (moulimukherjee) [reply src/adapter/src/catalog.rs]: updated

File: src/storage-types/src/connections/aws.rs
Comment (moulimukherjee) [reply src/storage-types/src/connections/aws.rs]: I made this part of the `impl AwsAssumeRole`

File: src/storage-types/src/connections/aws.rs
Comment (moulimukherjee) [reply src/storage-types/src/connections/aws.rs:304]: Actually I think I will add a new error type for SDK errors.

File: src/storage-types/src/connections/aws.rs
Comment (moulimukherjee) [reply src/storage-types/src/connections/aws.rs]: Updated. 
* I added implementation `AwsCredentials::load_shared_credentials_provider` and `AwsAssumeRole::load_shared_credentials_provider` instead. Because to load the sdk_config I will also need region and endpoint which are part of the parent `AwsConfig`.
* The public interface is still `AwsConfig::load_sdk_config(&self, connection_context: &ConnectionContext, connection_id: &GlobalId)`.
* Added a `AwsAssumeRole::dangerously_load_shared_credentials_provider_without_external_id` only called by the validate method

File: src/sql/src/plan/statement/ddl/connection.rs
Comment (moulimukherjee) [reply src/sql/src/plan/statement/ddl/connection.rs]: This is what I actually had before https://github.com/MaterializeInc/materialize/pull/23282#discussion_r1414155766

Comment (moulimukherjee) [reply src/sql/src/plan/statement/ddl/connection.rs]: I think I will revert back to using `sql_bail!` here. From the feedback it's making readability worse.

cc @jkosh44 @sploiselle Do you have better ideas here?

File: test/testdrive/mzcompose.py
Comment (moulimukherjee) [reply test/testdrive/mzcompose.py]: done

File: src/sql/src/plan/statement/ddl/connection.rs
Comment (sploiselle) [reply src/sql/src/plan/statement/ddl/connection.rs]: I really appreciate how you threaded the needle in making sure all of the possible states are covered here. One thing that might make the code a little more verbose would be to make the errors slightly less context sensitive––for example, you could error with all of the required options if any of them are missing. Users can do a little bit of deduction to figure out what they need.

Another thing that might help is to use the concrete types in these new error types and then rely on converting them to strings when displaying the error––this would get rid of the `to_string` noise, as well as coerce you to pare the errors down to only the minimal amount of context.

Glad to chat through any of this!

Comment (moulimukherjee) [reply src/sql/src/plan/statement/ddl/connection.rs]: > you could error with all of the required options if any of them are missing.

That's a good point. Let me do that, should clean up a bunch of match arms.

File: src/storage-types/src/connections/aws.rs
Comment (moulimukherjee) [reply src/storage-types/src/connections/aws.rs:351]: The validate method currently returns `Result<(), anyhow::Error>`, which eventually gets mapped to `AdapterError::Unstructured` with `SqlState::INTERNAL_ERROR` and no way to pass hint or detail for the error response. I will take a stab at refactoring that in a separate PR with a structured Error type so that hint and details can be passed properly. 

In this PR I am leaving things to be `anyhow!()`.

File: src/adapter/src/catalog/builtin_table_updates.rs
Comment (guswynn) [reply src/adapter/src/catalog/builtin_table_updates.rs]: resolved with the latest commit

**Comment by guswynn** [comment]
@benesch I resolved the borrow-checker issue, anything else we need before we can merge this?

File: src/adapter/src/catalog/builtin_table_updates.rs
Comment (benesch) [inline src/adapter/src/catalog/builtin_table_updates.rs:798]: Thanks, @guswynn!

**Comment by benesch** [comment]
Let me take a quick pass!

File: src/adapter/src/catalog/builtin_table_updates.rs
Comment (benesch) [inline src/adapter/src/catalog/builtin_table_updates.rs:788]: @guswynn I golfed this a little more to get rid of the holder entirely.

File: src/sql/src/plan/error.rs
Comment (benesch) [inline src/sql/src/plan/error.rs:236]: Why is this a separate type? We don't have any other example of splitting out `PlanError` within this crate.

File: src/sql/src/plan/error.rs
Comment (benesch) [inline src/sql/src/plan/error.rs:248]: These two options have AWS specific error messages but the variants don't indicate that they are AWS specific.

File: test/aws/mzcompose.py
Comment (benesch) [inline test/aws/mzcompose.py:28]: I renamed this composition to "aws" to reflect its new purpose as a general test of AWS-related functionality.

File: src/storage-types/src/connections/aws.rs
Comment (benesch) [inline src/storage-types/src/connections/aws.rs:370]: I substantially refactored the code in this file to have more comments, produce better error messages (e.g., there were several instances of `"error!"` with no additional context in user-facing error messages), and to reduce duplication.

**Review by benesch — COMMENTED** [review]
I pushed up a big batch of changes. This looks nearly mergeable. Outstanding blockers:

  1. [ ] Fixing the SQL option parsing enum
  2. [ ] Manual testing on staging

Other nice to haves:
  * [ ] Localstack tests for IAM (mis)configurations
  * [ ] Reference documentation for `mz_aws_connections`

**Review by philip-stoev — CHANGES_REQUESTED** [review]
Given that this introduces a new DDL syntax, it needs a misc/python/materialize/checks/all_checks/* test, to guard against Mz restarts and upgrades . Consider taking an existing test on secrets from that directory and cloning it a bit.

**Review by philip-stoev — DISMISSED** [review]
Given that this introduces a new DDL syntax, it needs a misc/python/materialize/checks/all_checks/* test, to guard against Mz restarts and upgrades . Consider taking an existing test on secrets from that directory and cloning it a bit.

**Comment by guswynn** [comment]
pushing a not-quite-finished commit to test something

File: src/storage-types/src/connections/aws.rs
Comment (benesch) [inline src/storage-types/src/connections/aws.rs:246]: And especially spooked that we need to override the region in the AssumeRole builder. This feels like something is wrong!

File: src/storage-types/src/connections/aws.rs
Comment (benesch) [inline src/storage-types/src/connections/aws.rs:363]: I'm a bit spooked that this was necessary

**Review by benesch — COMMENTED** [review]
@guswynn I can't figure out why the AWS SDK isn't automatically figuring out the region? Do you have a reference for that? 

It looks like maybe something got fixed in the latest version of the AWS SDK around regions and `AssumeRole`? Eyeballing https://github.com/smithy-lang/smithy-rs/pull/3014 and it seems like it could definitely help us.

File: src/storage-types/src/connections/aws.rs
Comment (guswynn) [reply src/storage-types/src/connections/aws.rs:246]: I believe the first one on line 226 is required, but I agree that its cursed (perhaps fixed by the pr you linked above?)

I followed https://github.com/MaterializeInc/materialize/blob/main/src/persist/src/s3.rs#L134-L137, and based my understanding on seeing this: https://docs.rs/aws-config/0.55.1/src/aws_config/lib.rs.html#563-567

Its possible the second `AssumeRoleProdider` doesn't need it, do you want me to try that?

Comment (guswynn) [reply src/storage-types/src/connections/aws.rs:247]: pushed a commit removing this second one to test this, after the build finishes ill test in staging

Comment (guswynn) [reply src/storage-types/src/connections/aws.rs:247]: rip, its required

Comment (benesch) [reply src/storage-types/src/connections/aws.rs:247]: You didn't bump the AWS SDK version did you?

Comment (guswynn) [reply src/storage-types/src/connections/aws.rs:247]: @benesch I didn't, because the next bump is from 0.x to 1.x and I don't want to block this pr on that change

**Comment by guswynn** [comment]
@philip-stoev pushed a commit with a basic check for these connections and their syntax

**Review by benesch — APPROVED** [review]

**Comment by moulimukherjee** [comment]
Thanks for all the efforts and getting this merged 🙏, this would have been deep in merge conflict land otherwise.

**Pull request merged**
